### ISSUE TITLE
reanimated 2.8.0 crashing android production app at Attempt to invoke virtual method 'java.util.Iterator java.util.HashSet.iterator()' on a null object reference

### ISSUE THREAD

**Issue by evelant** [software-mansion/react-native-reanimated#3265]
## Description

App hard crashes sometimes in Android production builds at `AnimationsManager.java:234`. This seems to happen mostly but not always when using `react-native-restart` to restart the app at runtime. Stack:

```
java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.Iterator java.util.HashSet.iterator()' on a null object reference
        at com.swmansion.reanimated.layoutReanimation.AnimationsManager.removeLeftovers(AnimationsManager.java:234)
        at com.swmansion.reanimated.layoutReanimation.AnimationsManager.lambda$scheduleCleaning$0$com-swmansion-reanimated-layoutReanimation-AnimationsManager(AnimationsManager.java:291)
        at com.swmansion.reanimated.layoutReanimation.AnimationsManager$$ExternalSyntheticLambda0.run(Unknown:4)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:27)
        at android.os.Looper.loop(Looper.java:237)
        at android.app.ActivityThread.main(ActivityThread.java:8167)
        at java.lang.reflect.Method.invoke(Method.java:-2)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
```

### Expected behavior
Reanimated should never cause native crashes

### Actual behavior & steps to reproduce
Unknown, use reanimated and attempt to use react-native-restart is the only semi-reliable trigger I've found but it doesn't always crash

## Snack or minimal code example
Unable, but I think it should be sufficient to place a null check at https://github.com/software-mansion/react-native-reanimated/blob/b4ee4ea9a1f246c461dd1819c6f3d48440a25756/android/src/main/java/com/swmansion/reanimated/layoutReanimation/AnimationsManager.java#L234 

## Package versions

| name                                 | version |
| ------------------------------------ | ------- |
| react-native                         |    0.68.2     |
| react-native-reanimated              |     2.8.0    |
| NodeJS                               |         |
| Xcode                                |         |
| Java                                 |         |
| Gradle                               |         |
| expo                                 |     45    |

## Affected platforms

- [ X ] Android
- [ ] iOS
- [ ] Web

**Comment by evelant** [comment]
Ah OK I think I see the source of the issue

https://github.com/software-mansion/react-native-reanimated/blob/b4ee4ea9a1f246c461dd1819c6f3d48440a25756/android/src/main/java/com/swmansion/reanimated/layoutReanimation/AnimationsManager.java#L284-L293

schedules the `removeLeftovers()` function to be run sometime later but

https://github.com/software-mansion/react-native-reanimated/blob/b4ee4ea9a1f246c461dd1819c6f3d48440a25756/android/src/main/java/com/swmansion/reanimated/layoutReanimation/AnimationsManager.java#L87-L100

sets `mToRemove` to null when the app is restarting, apparently before the `removeLeftovers` function is called. `removeLeftovers` then crashes trying to iterate over the now-null `mToRemove`

**Comment by FadiAboMsalam** [comment]
am also facing the same issue when changing our app language from english to arabic which requires change layout from LTR to RTL and restarting the app 

`
2022-05-23 16:40:16.058 23104-23104/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.*.*, PID: 23104
    java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.util.HashMap.containsKey(java.lang.Object)' on a null object reference
        at com.swmansion.reanimated.layoutReanimation.AnimationsManager.dfs(AnimationsManager.java:15)
        at com.swmansion.reanimated.layoutReanimation.AnimationsManager.removeLeftovers(AnimationsManager.java:9)
        at com.swmansion.reanimated.layoutReanimation.AnimationsManager.lambda$scheduleCleaning$0(AnimationsManager.java:3)
        at com.swmansion.reanimated.layoutReanimation.AnimationsManager.a(Unknown Source:0)
        at com.swmansion.reanimated.layoutReanimation.a.run(Unknown Source:4)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:1)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)`

**Comment by FadiAboMsalam** [comment]
Btw I tried to apply patch with #3266 PR to fix my issue and it didn't solve it

**Comment by jshinrock** [comment]
I experienced this error today. The fix in #3266 DID fix the issue in my case. Do you have plan to merge the PR?
